src/Microsoft.DotNet.Build.Tasks.Feed/BlobFeedAction.cs

```diff
@@ -57,133 +54,76 @@ public BlobFeedAction(string expectedFeedUrl, string accountKey, string indexDir
+                    Json = JObject.FromObject(sleetSettings)
+                };
 
-            await clientThrottle.WaitAsync();
-            string leaseId = string.Empty;
-            //this defines the lease for 15 seconds (max is 60) and 3000 milliseconds between requests
-            AzureBlobLease blobLease = new AzureBlobLease(feed.AccountName, feed.AccountKey, string.Empty, feed.ContainerName, uploadPath, Log, "15", "3000");
+                CloudStorageAccount storageAccount = CloudStorageAccount.Parse(source.ConnectionString);
+                AzureFileSystem fileSystem = new AzureFileSystem(new LocalCache(), new Uri(source.Path), new Uri(source.Path), storageAccount, source.Name, source.FeedSubPath);
 
-            bool isLeaseRequired = allowOverwrite && await feed.CheckIfBlobExists(uploadPath);
-
-            if (isLeaseRequired)
-            {
+                // In case the first Push attempt fails with an InvalidOperationException we Init the feed and retry the Push command once.
+                // Sleet internally retries 5 times on each package when the push operation fails so we don't need to retry ourselves.
                 try
                 {
-                    leaseId = blobLease.Acquire();
-                }
-                catch (Exception)
-                {
-                    Log.LogError($"Unable to obtain lease on {uploadPath}");
+                    bool result = await PushCommand.RunAsync(settings, fileSystem, items.ToList(), allowOverwrite, !allowOverwrite, new SleetLogger(Log));
+                    return result;
                 }
-            }
-            try
-            {
-                bool isExists = await feed.CheckIfBlobExists(uploadPath);
-                if (!isExists || allowOverwrite)
+                catch (InvalidOperationException ex) when (ex.Message.Contains("init"))
                 {
-                    Log.LogMessage($"Uploading {item} to {uploadPath}.");
-                    UploadClient uploadClient = new UploadClient(Log);
-                    await
-                        uploadClient.UploadBlockBlobAsync(
-                            ct,
-                            feed.AccountName,
-                            feed.AccountKey,
-                            feed.ContainerName,
-                            item,
-                            uploadPath,
-                            leaseId);
+                    Log.LogWarning($"Sub-feed {source.FeedSubPath} has not been initialized. Initializing now...");
+                    bool result = await InitCommand.RunAsync(settings, fileSystem, true, true, new SleetLogger(Log), CancellationToken);
 
+                    if (result)
+                    {
+                        Log.LogMessage($"Initializing sub-feed {source.FeedSubPath} succeeded!");
+                        result = await PushCommand.RunAsync(settings, fileSystem, items.ToList(), allowOverwrite, !allowOverwrite, new SleetLogger(Log));
```

Comment:
The reason I suggested async funcs was to avoid repeating all the RunAsync parameters and encapsulate the logging lines for maintainability. (My understanding is the `for` was there for this reason.)